File: Go/XServer/service_test.go
// Copyright (c) 2026 FrameworkEase Technologies. All rights reserved.
// Use of this source code is governed by a MIT-style
// license that can be found in the LICENSE file.

package XServer

import (
	"reflect"
	"testing"

	"github.com/frameworkease/Universal.X.gRPC/Go/internal/Proto/Greet"
	"github.com/stretchr/testify/assert"
)

func TestServe(t *testing.T) {
	type GreetFooService struct {
		Greet.UnimplementedFooServiceServer
	}

	type GreetBarService struct {
		Greet.UnimplementedBarServiceServer
	}

	defer func() {
		servicesMu.Lock()
		for i := len(services) - 1; i >= 0; i-- {
			entry := services[i].([]any)
			instance := entry[0]
			typ := reflect.ValueOf(instance).Type()
			if typ == reflect.TypeFor[*GreetFooService]() || typ == reflect.TypeOf((*GreetBarService)(nil)) {
				services = append(services[:i], services[i+1:]...)
			}
		}
		servicesMu.Unlock()
	}()

	assert.True(t, Serve[GreetFooService](&Greet.FooService_ServiceDesc), "Foo 服务应当注册成功。")
	assert.True(t, Serve[GreetBarService](&Greet.BarService_ServiceDesc), "Bar 服务应当注册成功。")
	assert.False(t, Serve[GreetFooService](&Greet.FooService_ServiceDesc), "Foo 服务只能注册一次。")
	assert.False(t, Serve[GreetBarService](&Greet.BarService_ServiceDesc), "Bar 服务只能注册一次。")

	servicesMu.RLock()
	hasFooService := false
	hasBarService := false
	for _, service := range services {
		entry := service.([]any)
		instance := entry[0]
		typ := reflect.ValueOf(instance).Type()
		if typ == reflect.TypeFor[*GreetFooService]() {
			hasFooService = true
		}
		if typ == reflect.TypeFor[*GreetBarService]() {
			hasBarService = true
		}
	}
	servicesMu.RUnlock()

	assert.True(t, hasFooService, "服务列表应当包含 Foo 服务。")
	assert.True(t, hasBarService, "服务列表应当包含 Bar 服务。")
}

File: CSharp/tests~/Assets/Tests/Runtime/Scripts/XServer/service_test.cs
// Copyright (c) 2026 FrameworkEase Technologies. All rights reserved.
// Use of this source code is governed by a MIT-style
// license that can be found in the LICENSE file.

#if !UNITY_5_3_OR_NEWER
using FrameworkEase.Universal.X.gRPC;
using FrameworkEase.Universal.X.gRPC.Tests.Proto.Greet;
using NUnit.Framework;

public partial class TestXServer
{
    [Test]
    public void Serve()
    {
        try
        {
            Assert.That(XServer.Serve(typeof(FooService)), Is.True, "Foo 服务应当注册成功。");
            Assert.That(XServer.Serve(typeof(BarService)), Is.True, "Bar 服务应当注册成功。");
            Assert.That(XServer.Serve(typeof(FooService)), Is.False, "Foo 服务只能注册一次。");
            Assert.That(XServer.Serve(typeof(BarService)), Is.False, "Bar 服务只能注册一次。");
            Assert.That(XServer.Services, Does.Contain(typeof(FooService)), "服务列表应当包含 Foo 服务。");
            Assert.That(XServer.Services, Does.Contain(typeof(BarService)), "服务列表应当包含 Bar 服务。");
        }
        finally { XServer.Services.RemoveAll(ele => ele == typeof(FooService) || ele == typeof(BarService)); }
    }
}
#endif
